### ISSUE TITLE
How to customize mapping of types implementing IEnumerable to DynamicObject?

### ISSUE THREAD

**Issue by azabluda** [6bee/aqua-core#7]
Sometimes we need to map types which implement `IEnumerable` but also contain other properties. The standard example is `IGrouping<TKey, TElement>` which implements `IEnumerable<TElement>` but also contains the `Key` property.
```c#
public interface IGrouping<out TKey, out TElement> : IEnumerable<TElement>, IEnumerable
{
    TKey Key { get; }
}
```
Class `DynamicObject` provides a virtual method `MapToDynamicObjectGraph` which I think I am supposed to override, f.ex. the following way
```c#
class CustomDynamicObjectMapper : DynamicObjectMapper
{
    protected override DynamicObject MapToDynamicObjectGraph(object obj, Func<Type, bool> setTypeInformation)
    {
        if (obj != null && obj.GetType().Implements(typeof(IGrouping<,>)))
        {
            object mappedGrouping = typeof(CustomDynamicObjectMapper)
                .GetTypeInfo()
                .GetDeclaredMethod(nameof(MapGrouping))
                .MakeGenericMethod(obj.GetType().GenericTypeArguments)
                .Invoke(this, new[] { obj, setTypeInformation });
            return (DynamicObject)mappedGrouping;
        }

        return base.MapToDynamicObjectGraph(obj, setTypeInformation);
    }

    private DynamicObject MapGrouping<TKey, TElement>(IGrouping<TKey, TElement> grouping, Func<Type, bool> setTypeInformation)
    {
        var mappedGrouping = new DynamicObject(typeof(IGrouping<TKey, TElement>));
        mappedGrouping.Add("Key", this.MapToDynamicObjectGraph(grouping.Key, setTypeInformation));
        mappedGrouping.Add("Elements", this.MapCollection(grouping, setTypeInformation).ToList());
        return mappedGrouping;
    }
}
```
and it seems to return me exactly what I want
```c#
IGrouping<string, int> grouping = Enumerable.Range(1, 5).GroupBy(x => "Hello", x => x).Single();
DynamicObject dynamicGrouping = new CustomDynamicObjectMapper().MapObject(grouping);
```
Unfortunately the overridden method is NOT called when mapping an **object containing an `IGrouping` property**, which makes me believe I chose the wrong way of custom type mapping. What could you advise instead?

My example

```c#
public class When_mapping_object_from_object_with_igrouping_members
{
    class CustomClass
    {
        public IGrouping<string, int> Grouping { get; set; }
    }

    class CustomDynamicObjectMapper : DynamicObjectMapper
    {
        protected override DynamicObject MapToDynamicObjectGraph(object obj, Func<Type, bool> setTypeInformation)
        {
            if (obj != null && obj.GetType().Implements(typeof(IGrouping<,>)))
            {
                object mappedGrouping = typeof(CustomDynamicObjectMapper)
                    .GetTypeInfo()
                    .GetDeclaredMethod(nameof(MapGrouping))
                    .MakeGenericMethod(obj.GetType().GenericTypeArguments)
                    .Invoke(this, new[] { obj, setTypeInformation });
                return (DynamicObject)mappedGrouping;
            }

            return base.MapToDynamicObjectGraph(obj, setTypeInformation);
        }

        private DynamicObject MapGrouping<TKey, TElement>(IGrouping<TKey, TElement> grouping, Func<Type, bool> setTypeInformation)
        {
            var mappedGrouping = new DynamicObject(typeof(IGrouping<TKey, TElement>));
            mappedGrouping.Add("Key", this.MapToDynamicObjectGraph(grouping.Key, setTypeInformation));
            mappedGrouping.Add("Elements", this.MapCollection(grouping, setTypeInformation).ToList());
            return mappedGrouping;
        }
    }

    DynamicObject dynamicGrouping;
    DynamicObject dynamicObject;

    public When_mapping_object_from_object_with_igrouping_members()
    {
        IGrouping<string, int> grouping = Enumerable.Range(1, 5).GroupBy(x => "Hello", x => x).Single();
        var source = new CustomClass
        {
            Grouping = grouping
        };
        dynamicGrouping = new CustomDynamicObjectMapper().MapObject(grouping);
        dynamicObject = new CustomDynamicObjectMapper().MapObject(source);
    }

    [Fact]
    public void Dynamic_grouping_type_should_be_igrouping()
    {
        dynamicGrouping.Type.Type.Implements(typeof(IGrouping<,>)).ShouldBeTrue();
    }

    [Fact]
    public void Dynamic_object_grouping_should_be_mapped()
    {
        dynamicObject["Grouping"].ShouldBeOfType<DynamicObject>();
    }

    [Fact]
    public void Dynamic_object_grouping_type_should_be_igrouping()
    {
        ((DynamicObject)dynamicObject["Grouping"]).Type.Type.Implements(typeof(IGrouping<,>)).ShouldBeTrue();
    }
}
```
Both asserts on `dynamicObject` fail.

**Comment by 6bee** [comment]
Your attempt looks perfectly fine to me. 
The flaw is in the base mapper, treating all IEnumerable types as collections. Probably we need to add another virtual method to decide whether to map as collection or single object e.g. bool ShouldMapToDynamicObject(IEnumerable)

**Comment by azabluda** [comment]
That should work, though I would find it more natural to have just one virtual method where I either provide my custom mapping (no matter IEnumerable or not) or pass it on to your base implementation which handles objects and collections the usual way.

**Comment by 6bee** [comment]
I see your point, but it makes a difference whether MapToDynamicObjectGraph returns a single object or a collection. That's why ShouldMapToDynamicObject would first check whether to pass the entire IEnumerable to MapToDynamicObjectGraph, to map it into a single DynamicObject, or rather pass element by element which is the default behaviour.

**Comment by azabluda** [comment]
If the distinction is so important at that stage, then I fully agree with your proposal.

**Comment by 6bee** [comment]
resolved in commit be1c7a1

Now you can override `ShouldMapToDynamicObject` in subtyped `DynamicObjectMapper` and return true for collections you want to be passed to `MapToDynamicObjectGraph` for mapping.

sample:
```
class CustomDynamicObjectMapper : DynamicObjectMapper
{
  protected override DynamicObject MapToDynamicObjectGraph(object obj, Func<Type, bool> setTypeInformation)
  {
    if (obj != null && obj.GetType().Implements(typeof(IGrouping<,>)))
    {      
      return // custom mapping goes here
    }

    return base.MapToDynamicObjectGraph(obj, setTypeInformation);
  }

  protected override bool ShouldMapToDynamicObject(IEnumerable collection)
  {
    return collection.GetType().Implements(typeof(IGrouping<,>));
  }
}
```

**Comment by azabluda** [comment]
Looks neat :+1:
